### PR TITLE
[SPARK-30102][ML][PYSPARK] GMM supports instance weighting

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
supports instance weighting in GMM

### Why are the changes needed?
ML should support instance weighting

### Does this PR introduce any user-facing change?
yes, a new param `weightCol` is exposed

### How was this patch tested?
added testsuits

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/clustering/GaussianMixtureSuite.scala [hunk 1]
```diff
@@ -267,6 +268,18 @@ class GaussianMixtureSuite extends MLTest with DefaultReadWriteTest {
     assert(trueLikelihood ~== floatLikelihood absTol 1e-6)
   }
 
+  test("GMM support instance weighting") {
```
Comment (zhengruifeng) [inline mllib/src/test/scala/org/apache/spark/ml/clustering/GaussianMixtureSuite.scala]: The final model is highly sensitive to the initlization, which is computed in `initRandom` and highly related to the data distribution & partition.

**Review by srowen — COMMENTED** [review]
I don't suppose there's any clean way to avoid the added computation for weights when the weights are all 1? looks hard. I'm OK with it

**Comment by zhengruifeng** [comment]
There seems something wrong in the py doctests.
1, I manually test some scala cases/examples between 2.4.4 and this PR, the results are expected.
2, I manually test the py doctest in 2.4.4 and the result is different from current expected value:
image
3, I manually test the py doctest in 2.4.4 and the result the same as 2.4.4:
image

I think I need to look into this.

**Comment by zhengruifeng** [comment]
I also test it under python2.7 & python3.7, and the results are the same.

**Comment by zhengruifeng** [comment]
It really took me some days to look into the test failure.

1, In 2.4.4, I can not reproduce the doctests:
```python
    >>> summary.clusterSizes
    [2, 2, 2]
    >>> summary.logLikelihood
    8.14636...
```
until I explictly set numPartition=2, like this `df = spark.createDataFrame(sc.parallelize(data, 2), ["features"])`.
That is because existing `df = spark.createDataFrame(data, ["features"])` will create a df with 12 partitions, and GMM is highly sensitive to the intialization.
It is also weird to me that `spark.createDataFrame` will create a df with 6 partitions in the scala side.
My latop has a 8850 cpu with 6cores and 12threads.

2, After using `df = spark.createDataFrame(sc.parallelize(data, 2), ["features"])`, I can reproduce the results in 2.4.4. However, the doctests still fail, I log the optimization metric `logLikelihood` after each iteration and find that it seems a sudden numeric change.

----
Iteration|0|1|2|3|4|5|6|7|8|9|
Master|-13.306466494963615|-0.4307654468425961|0.49157579336057605|2.234212048899172|6.125367537295512|11.27762326533469|35.232285502171976|10.028821186214191|23.693392686726106|8.146360246481793|
This PR|-13.306466494963615|-0.430765446842597|0.4915757933605755|2.234212048899182|6.125367537295558|11.277623265335476|35.229680601767065|46.33491773124833|57.694248782061024|26.193922336279954|
---

The metrics are near before iter-7, but some sudden numeric change happened in iter-7. But I think it is acceptable since the internal computation is complex.
Moreover, current convergence check `math.abs(logLikelihood - logLikelihoodPrev) > $(tol)` do not work when optimization objective meet a big hit. Like `logLikelihood` drop from 35.232285502171976 to 10.028821186214191 in iter-7.

So I think I need to:
1, change the df generation logic with explictly set numpartition; (current `createDataFrame` do not support this input, I need to create a rdd first)
2, change the result in the doctest (I tend to set `MaxIter=5` and result=11.27)
3, change the convergence check and avoid big drop in optimzation metric(maybe in another PR and check other algs in it)

@srowen @huaxingao  How do you think about it?

**Comment by zhengruifeng** [comment]
> I don't suppose there's any clean way to avoid the added computation for weights when the weights are all 1? looks hard. I'm OK with it

An approach is to have two method: 
1, def add(instance: Vector, weight: Double): this.type = {
if(weight==1) add(instance) else ...
}
2, def add(instance: Vector): this.type

the second method do not include any extra multiplication.

**Comment by huaxingao** [comment]
I guess instead of changing maxIter=5 and compare the logLikelihood at iteration 5, maybe use a much bigger maxIter  so it will converge. Compare the logLikelihood at convergence.

It puzzled me why the logLikelihoods from iteration 7 are so different from the logLikelihoods computed using the original code. Weight is not set in the python doctest and it uses default 1.0. So in theory, this should behave exact the same as the original code,  the logLikelihood at each iteration should be very similar as the logLikelihood computed using the original code, right?

I tried both the original code and the code with changes, they start to have different logLikelihood at iteration 7, but both of them converge at iteration 25, with the same logLikelihood 65.02945125241477.

I agree that we probably need to change the current convergence check. Seems to me that we also need to compare the logLikelihood difference to the previous difference. The difference should be smaller and smaller and eventually converge. However, I tested with the current code, the logLikelihood differences are not getting smaller consistently. 

| iteration  | logLikelihoodPrev | logLikelihood |diff |
| --------  | ----------------- |-------------- | -- |
| 20  | 36.402816949681664 | 36.55682231506764 | 0.1540053653859772|
| 16  | 36.55682231506764 | 36.75888971475007 | 0.20206739968242715|
| 17  | 36.75888971475007 | 37.581643170088086 | 0.8227534553380167|
| 18  | 37.581643170088086 | 6.674670202869423 | 30.906972967218664|
| 19  | 6.674670202869423 | 10.601046748584544 | 3.9263765457151205|
| 20  | 10.601046748584544 | 39.71941181091317 | 29.11836506232863|
| 21  | 39.71941181091317 | 49.2147989416624 | 9.49538713074923|
| 22  | 49.2147989416624 | 76.11383657713708 | 26.899037635474677|
| 23  | 76.11383657713708 | 71.28238165058754 | 4.83145492654954|
| 24  | 71.28238165058754 | 65.02945125241477 | 6.252930398172765|
| 25  | 65.02945125241477 | 65.02945125241477 | 0.0|

**Comment by zhengruifeng** [comment]
@huaxingao Thanks a lot for looking into this! 
Yes, it is reasonable to set MaxIter=25 for a stable result, and I still need to look for the reason for the divergence at iter-7.
And in theory the loglikehood should be increasing, we should check the training procedure.

**Comment by srowen** [comment]
@zhengruifeng does this need more investigation?

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
@huaxingao I found that the difference come from the method to compute var `sumWeights`.
In master, it keep the weights for each clustering, and get the sum by `sums.weights.sum`,
while in previous commit, I use a seperate var to keep the sum. The difference is quite tiny, but it unfortunately cause a sudden divergence at iter-7 (although the two impl finally convergen to the same result).
I revert the computation of `sumWeights` and now the doctest works fine.

@srowen I think this PR should be OK to merge. However, as disscussed above, we can see that the convergence of GMM is not stable, the loglikelihood may even drop sharply during the training procedure (which should not happen in theory). I think we need to make GMM more numerical stable in the future, but as to supporting instance weighting, I think current PR is OK.

File: mllib/src/main/scala/org/apache/spark/ml/util/MetadataUtils.scala [hunk 2]
```diff
@@ -41,6 +41,24 @@ private[spark] object MetadataUtils {
     }
   }
 
+  /**
+   * Examine a schema to identify the number of features in a vector column.
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/util/MetadataUtils.scala]: The same method is also added in PR for RobustScaler, I tend to merge that first, and rebase this PR.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala [hunk 3]
```diff
@@ -526,11 +546,14 @@ class GaussianMixture @Since("2.0.0") (
        */
       val cov = {
         val ss = new DenseVector(new Array[Double](numFeatures)).asBreeze
-        slice.foreach(xi => ss += (xi.asBreeze - mean.asBreeze) ^:^ 2.0)
+        var j = 0
+        while (j < numSamples) {
+          ss += ((sampleSlice(j).asBreeze - mean.asBreeze) ^:^ 2.0) * weightSlice(j)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: There's a similar bit of code I had to change recently to compile in Scala 2.13. You don't have to address it here, but, I think that using `val v: BV[Double] = sampleSlice(j).asBreeze - mean.asBreeze` first fixes it

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala [hunk 4]
```diff
@@ -500,21 +511,30 @@ class GaussianMixture @Since("2.0.0") (
...
-      val slice = samples.view(i * numSamples, (i + 1) * numSamples)
+    val (samples, sampleWeights) = instances
+      .takeSample(withReplacement = true, numClusters * numSamples, $(seed))
+      .unzip
+
+    val weights = new Array[Double](numClusters)
+    val weightSum = sampleWeights.sum
+
+    val gaussians = Array.tabulate(numClusters) { i =>
+      val sampleSlice = samples.view(i * numSamples, (i + 1) * numSamples)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: Nit: you could factor out `i * numSamples` here and use it 4 times. Might be very slightly tidier

**Review by srowen — COMMENTED** [review]
Looking OK to me

**Comment by zhengruifeng** [comment]
retest this please

**Review by srowen — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Merged to master, thanks @srowen @huaxingao for reviewing!

**Pull request closed without merging**